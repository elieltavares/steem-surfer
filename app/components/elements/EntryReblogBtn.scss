@import '../../style/mixins';
@import '../../style/colors';

@keyframes reblogging-color-change {
  0% {
    color: $dark-sky-blue;
  }
  25% {
    color: transparent;
  }
  50% {
    color: $dark-sky-blue;
  }
  75% {
    color: transparent;
  }
  100% {
    color: $dark-sky-blue;
  }
}

.reblog-btn {
  &.processing {
    animation: reblogging-color-change 5s infinite;
    pointer-events: none;

    .mi {
      opacity: 1 !important;
    }
  }
}
