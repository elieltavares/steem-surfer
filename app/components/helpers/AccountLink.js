import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { getAccount } from '../../backend/steem-client';

export const makePath = username => `/@${username}`;

class AccountLink extends Component {
  goProfile = async () => {
    const { username, history, actions, onClick, afterClick } = this.props;

    onClick();

    let { accountData } = this.props;
    const { setVisitingAccount } = actions;

    if (!accountData) {
      try {
        accountData = await getAccount(username);
      } catch (err) {
        accountData = null;
      }
    }

    if (accountData) {
      setVisitingAccount(accountData);
    }

    history.push(makePath(username));

    afterClick();
  };

  render() {
    const { children } = this.props;

    return React.cloneElement(children, {
      onClick: this.goProfile
    });
  }
}

AccountLink.defaultProps = {
  accountData: null,
  onClick: () => {},
  afterClick: () => {}
};

AccountLink.propTypes = {
  children: PropTypes.element.isRequired,
  username: PropTypes.string.isRequired,
  history: PropTypes.instanceOf(Object).isRequired,
  accountData: PropTypes.instanceOf(Object),
  onClick: PropTypes.func,
  afterClick: PropTypes.func,
  actions: PropTypes.shape({
    setVisitingAccount: PropTypes.func.isRequired
  }).isRequired
};

export default AccountLink;
