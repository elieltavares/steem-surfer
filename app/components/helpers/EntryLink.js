import React, { Component } from 'react';
import PropTypes from 'prop-types';

import { getContent } from '../../backend/steem-client';

export const makePath = (category, author, permlink, toReplies = false) =>
  `/${category}/@${author}/${permlink}${toReplies ? '#replies' : ''}`;

class EntryLink extends Component {
  goEntry = async () => {
    let { entry } = this.props;
    const { history, actions, toReplies, markAsRead } = this.props;

    if (markAsRead) {
      markAsRead();
    }

    if (!entry) {
      const { author, permlink } = this.props;
      entry = await getContent(author, permlink);
    }

    if (entry) {
      const { category, author, permlink } = entry;
      actions.setVisitingEntry(entry);
      history.push(makePath(category, author, permlink, toReplies));
    }
  };

  render() {
    const { children } = this.props;

    return React.cloneElement(children, {
      onClick: this.goEntry
    });
  }
}

EntryLink.defaultProps = {
  entry: null,
  toReplies: false,
  markAsRead: null
};

EntryLink.propTypes = {
  children: PropTypes.element.isRequired,
  history: PropTypes.instanceOf(Object).isRequired,
  entry: PropTypes.shape({
    category: PropTypes.string.isRequired,
    author: PropTypes.string.isRequired,
    permlink: PropTypes.string.isRequired
  }),
  author: PropTypes.string.isRequired,
  permlink: PropTypes.string.isRequired,
  toReplies: PropTypes.bool,
  actions: PropTypes.shape({
    setVisitingEntry: PropTypes.func.isRequired
  }).isRequired,
  markAsRead: PropTypes.func
};

export default EntryLink;
