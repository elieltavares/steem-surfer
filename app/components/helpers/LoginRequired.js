import React, { Component, Fragment } from 'react';
import PropTypes from 'prop-types';

import { Modal } from 'antd';
import { FormattedMessage, FormattedHTMLMessage } from 'react-intl';

import Login from '../dialogs/Login';

export const checkLogin = (activeAccount, requiredKeys) => {
  // Steem connect login.
  if (activeAccount && activeAccount.type === 'sc') {
    return true;
  }

  // Traditional login. Check all required keys exists in user keys.
  if (
    activeAccount &&
    activeAccount.type === 's' &&
    requiredKeys.every(e => activeAccount.keys[e])
  ) {
    return true;
  }

  return false;
};

class LoginRequired extends Component {
  constructor(props) {
    super(props);

    this.state = {
      loginModalVisible: false
    };
  }

  showLoginModal = () => {
    this.setState({
      loginModalVisible: true
    });

    setTimeout(() => {
      const { onDialogOpen } = this.props;
      onDialogOpen();
    }, 200);
  };

  onLoginModalCancel = () => {
    this.setState({
      loginModalVisible: false
    });
  };

  onLoginSuccess = () => {
    this.setState({
      loginModalVisible: false
    });

    setTimeout(() => {
      if (!this.check()) {
        this.showLoginModal();
      }
    }, 400);
  };

  afterModalClose = () => {
    const { onDialogClose } = this.props;
    onDialogClose();
  };

  check = () => {
    const { activeAccount, requiredKeys } = this.props;

    return checkLogin(activeAccount, requiredKeys);
  };

  render() {
    const { children } = this.props;

    if (this.check()) {
      return children;
    }

    const { activeAccount, requiredKeys } = this.props;
    const { loginModalVisible } = this.state;

    // Default message
    let loginMsg = <FormattedMessage id="login-required.default" />;

    if (activeAccount) {
      // More specific, key based message
      loginMsg =
        requiredKeys.length > 1 ? (
          <FormattedHTMLMessage
            id="login-required.keys-required"
            values={{ keys: requiredKeys }}
          />
        ) : (
          <FormattedHTMLMessage
            id="login-required.key-required"
            values={{ key: requiredKeys }}
          />
        );
    }

    // Clone child element rewriting onclick event
    const newChild = React.cloneElement(children, {
      onClick: () => {
        this.showLoginModal();
      }
    });

    return (
      <Fragment>
        {newChild}

        <Modal
          visible={loginModalVisible}
          onCancel={this.onLoginModalCancel}
          footer={false}
          width="500px"
          closable={false}
          destroyOnClose
          centered
          afterClose={this.afterModalClose}
        >
          <Login
            {...this.props}
            onSuccess={this.onLoginSuccess}
            loginMsg={loginMsg}
          />
        </Modal>
      </Fragment>
    );
  }
}

LoginRequired.defaultProps = {
  activeAccount: null,
  onDialogOpen: () => {},
  onDialogClose: () => {},
  requiredKeys: []
};

LoginRequired.propTypes = {
  children: PropTypes.element.isRequired,
  activeAccount: PropTypes.instanceOf(Object),
  requiredKeys: PropTypes.arrayOf(PropTypes.string),
  onDialogOpen: PropTypes.func,
  onDialogClose: PropTypes.func
};

export default LoginRequired;
