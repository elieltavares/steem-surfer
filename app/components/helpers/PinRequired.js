import React, { Component, Fragment } from 'react';
import { Modal } from 'antd';

import PropTypes from 'prop-types';
import PinConfirm from '../dialogs/PinConfirm';

class PinRequired extends Component {
  constructor(props) {
    super(props);

    this.state = {
      show: false
    };
  }

  onPinInvalidated = () => {
    this.setState({ show: false });
  };

  onConfirmPinSuccess = value => {
    const { onSuccess } = this.props;
    this.setState({ show: false }, () => {
      onSuccess(value);
    });
  };

  cancelFn = () => {
    this.setState({ show: false });
  };

  render() {
    const { children } = this.props;
    const { show } = this.state;

    // Clone child element rewriting onclick event
    const newChild = React.cloneElement(children, {
      onClick: () => {
        this.setState({ show: true });
      }
    });

    return (
      <Fragment>
        {show && (
          <Modal
            footer={null}
            closable
            onCancel={this.cancelFn}
            keyboard
            visible
            width="500px"
            centered
            destroyOnClose
          >
            <PinConfirm
              {...this.props}
              onSuccess={this.onConfirmPinSuccess}
              onInvalidate={this.onPinInvalidated}
            />
          </Modal>
        )}
        {newChild}
      </Fragment>
    );
  }
}

PinRequired.defaultProps = {
  onSuccess: () => {}
};

PinRequired.propTypes = {
  children: PropTypes.element.isRequired,
  history: PropTypes.instanceOf(Object).isRequired,
  onSuccess: PropTypes.func
};

export default PinRequired;
