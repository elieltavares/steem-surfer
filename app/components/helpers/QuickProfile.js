/*
eslint-disable import/no-cycle
*/

import React, { Component, Fragment } from 'react';
import PropTypes from 'prop-types';
import { Drawer, message } from 'antd';
import { FormattedMessage, FormattedNumber } from 'react-intl';

import { proxifyImageSrc } from '@esteemapp/esteem-render-helpers';

import Tooltip from '../common/Tooltip';

import {
  getAccount,
  getDiscussions,
  getFollowCount
} from '../../backend/steem-client';
import authorReputation from '../../utils/author-reputation';

import UserAvatar from '../elements/UserAvatar';
import EntryListLoadingItem from '../elements/EntryListLoadingItem';
import LinearProgress from '../common/LinearProgress';
import EntryListItem from '../elements/EntryListItem';
import FollowControls from '../elements/FollowControls';
import AccountLink from './AccountLink';
import coverFallbackDay from '../../img/cover-fallback-day.png';
import coverFallbackNight from '../../img/cover-fallback-night.png';

class QuickProfile extends Component {
  constructor(props) {
    super(props);

    this.state = {
      visible: false,
      active: false,
      profile: {
        name: ' ',
        about: ' ',
        coverImage: null,
        postCount: 0
      },
      follows: {
        followerCount: null,
        followingCount: null
      },
      loadingEntries: true,
      entries: [],
      accountData: null
    };
  }

  load = async () => {
    const { username, intl } = this.props;

    const account = await getAccount(username);

    this.setState({ accountData: account });

    let accountProfile;
    try {
      accountProfile = JSON.parse(account.json_metadata).profile;
    } catch (err) {
      accountProfile = {};
    }

    if (accountProfile) {
      const name = accountProfile.name || null;
      const about = accountProfile.about || null;
      const postCount = account.post_count;
      const coverImage = accountProfile.cover_image || null;

      this.setState({ profile: { name, about, coverImage, postCount } });
    }

    let follow;
    try {
      follow = await getFollowCount(username);
    } catch (err) {
      message.error(
        intl.formatMessage({ id: 'quick-profile.fetch-error-profile' })
      );
    }

    if (follow) {
      const followerCount = follow.follower_count;
      const followingCount = follow.following_count;

      this.setState({ follows: { followerCount, followingCount } });
    }

    let entries;
    try {
      entries = await getDiscussions('blog', {
        tag: username,
        limit: 7,
        start_author: undefined,
        start_permlink: undefined
      });
    } catch (err) {
      message.error(
        intl.formatMessage({ id: 'quick-profile.fetch-error-content' })
      );
    }

    if (entries) {
      this.setState({ entries, loadingEntries: false });
    }
  };

  show = () => {
    this.setState({ active: true, visible: true });
    this.load();
  };

  hide = () => {
    this.setState({ visible: false });
    setTimeout(() => {
      this.setState({ active: false });
    }, 500);
  };

  goSection = section => {
    const { history, username } = this.props;
    const u = `/@${username}/${section}`;
    history.push(u);
  };

  render() {
    const { children, username, global, intl } = this.props;
    let { reputation } = this.props;
    const {
      visible,
      active,
      profile,
      follows,
      entries,
      loadingEntries,
      accountData
    } = this.state;

    if (accountData) ({ reputation } = accountData);

    const newChild = React.cloneElement(children, {
      onClick: () => {
        this.show();
      }
    });

    const followersMsg =
      follows.followerCount === null ? (
        '--'
      ) : (
        <FormattedNumber value={follows.followerCount} />
      );
    const postCountMsg =
      profile.postCount === null ? (
        '--'
      ) : (
        <FormattedNumber value={profile.postCount} />
      );
    const followingMsg =
      follows.followingCount === null ? (
        '--'
      ) : (
        <FormattedNumber value={follows.followingCount} />
      );

    const bgImage =
      (profile.coverImage && proxifyImageSrc(profile.coverImage)) ||
      (global.theme === 'day' ? coverFallbackDay : coverFallbackNight);

    return (
      <Fragment>
        {newChild}

        {active && (
          <Drawer
            placement="right"
            closable={false}
            onClose={this.hide}
            visible={visible}
            width="640px"
          >
            <div
              className={`quick-profile-content ${
                loadingEntries ? 'loading' : ''
              } `}
            >
              <div className="profile-area">
                <div
                  className="account-cover"
                  style={{ backgroundImage: `url('${bgImage}')` }}
                >
                  <div className="profile-avatar">
                    <UserAvatar {...this.props} user={username} size="large" />
                    <div className="reputation">
                      {authorReputation(reputation)}
                    </div>
                  </div>
                  <div className="follow-btn-holder">
                    <FollowControls {...this.props} targetUsername={username} />
                  </div>
                </div>
                <AccountLink
                  {...this.props}
                  username={username}
                  accountData={accountData}
                >
                  <div className="username">{username}</div>
                </AccountLink>
                <div className="about">{profile.about}</div>
                <div className="numbers">
                  <Tooltip
                    title={intl.formatMessage({
                      id: 'quick-profile.post-count'
                    })}
                  >
                    <span className="post-count">
                      <i className="mi">list</i>
                      {postCountMsg}
                    </span>
                  </Tooltip>
                  <Tooltip
                    title={intl.formatMessage({
                      id: 'quick-profile.followers'
                    })}
                  >
                    <span className="followers">
                      <i className="mi">people</i>
                      {followersMsg}
                    </span>
                  </Tooltip>
                  <Tooltip
                    title={intl.formatMessage({
                      id: 'quick-profile.following'
                    })}
                  >
                    <span className="following">
                      <i className="mi">person_add</i>
                      {followingMsg}
                    </span>
                  </Tooltip>
                </div>
              </div>

              <div className="account-menu">
                <div className="account-menu-items">
                  <a
                    role="none"
                    className="menu-item selected-item"
                    onClick={() => {
                      this.goSection('blog');
                    }}
                  >
                    <FormattedMessage id="quick-profile.section-blog" />
                  </a>
                  <a
                    role="none"
                    className="menu-item"
                    onClick={() => {
                      this.goSection('comments');
                    }}
                  >
                    <FormattedMessage id="quick-profile.section-comments" />
                  </a>
                  <a
                    role="none"
                    className="menu-item"
                    onClick={() => {
                      this.goSection('replies');
                    }}
                  >
                    <FormattedMessage id="quick-profile.section-replies" />
                  </a>
                  <span className="separator-item" />
                  <a
                    role="none"
                    className="menu-item"
                    onClick={() => {
                      this.goSection('points');
                    }}
                  >
                    <FormattedMessage id="quick-profile.section-points" />
                  </a>
                  <a
                    role="none"
                    className="menu-item"
                    onClick={() => {
                      this.goSection('wallet');
                    }}
                  >
                    <FormattedMessage id="quick-profile.section-wallet" />
                  </a>
                </div>
              </div>

              {loadingEntries && (
                <Fragment>
                  <LinearProgress />
                  <div className="entries">
                    <EntryListLoadingItem />
                  </div>
                </Fragment>
              )}

              {!loadingEntries && (
                <div className="entries">
                  {entries.map(d => (
                    <EntryListItem
                      key={`${d.author}-${d.permlink}`}
                      {...this.props}
                      entry={d}
                      asAuthor={username}
                      inDrawer
                    />
                  ))}
                </div>
              )}
            </div>
          </Drawer>
        )}
      </Fragment>
    );
  }
}

QuickProfile.defaultProps = {};

QuickProfile.propTypes = {
  children: PropTypes.element.isRequired,
  username: PropTypes.string.isRequired,
  reputation: PropTypes.oneOfType([PropTypes.string, PropTypes.number])
    .isRequired,
  history: PropTypes.instanceOf(Object).isRequired,
  global: PropTypes.shape({
    theme: PropTypes.string.isRequired
  }).isRequired,
  intl: PropTypes.instanceOf(Object).isRequired
};

export default QuickProfile;
