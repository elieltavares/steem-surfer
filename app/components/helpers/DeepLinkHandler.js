import React from 'react';
import PropTypes from 'prop-types';

import filters from '../../constants/filters.json';

import { makePath as makePathTag } from './TagLink';
import { makePath as makePathAccount } from './AccountLink';
import { makePath as makePathEntry } from './EntryLink';

export const deepLink2Obj = url => {
  let urlPart = url.split('://')[1];

  // remove last char if /
  if (urlPart.endsWith('/')) {
    urlPart = urlPart.substring(0, urlPart.length - 1);
  }

  const parts = urlPart.split('/');

  // witnesses
  if (parts.length === 1 && parts[0] === 'witnesses') {
    return { type: 'witnesses' };
  }

  // sps
  if (parts.length === 1 && parts[0] === 'sps') {
    return { type: 'sps' };
  }

  // filter
  if (parts.length === 1 && filters.includes(parts[0])) {
    return { type: 'filter', filter: parts[0] };
  }

  // filter with tag
  if (parts.length === 2 && filters.includes(parts[0])) {
    return { type: 'filter-tag', filter: parts[0], tag: parts[1] };
  }

  // account
  if (parts.length === 1 && parts[0].startsWith('@')) {
    return { type: 'account', account: parts[0].replace('@', '') };
  }

  // post without tag
  if (parts.length === 2 && parts[0].startsWith('@')) {
    return {
      type: 'post',
      cat: 'esteem',
      author: parts[0].replace('@', ''),
      permlink: parts[1]
    };
  }

  // post with tag
  if (parts.length === 3 && parts[1].startsWith('@')) {
    return {
      type: 'post',
      cat: parts[0],
      author: parts[1].replace('@', ''),
      permlink: parts[2]
    };
  }
};

class DeepLinkHandler extends React.Component {
  componentDidMount() {
    window.addEventListener('deep-link', this.handler);
  }

  componentWillUnmount() {
    window.removeEventListener('deep-link', this.handler);
  }

  handler = e => {
    const { history } = this.props;
    const { url } = e.detail;
    const obj = deepLink2Obj(url);

    let path;

    switch (obj.type) {
      case 'filter':
        path = `/${obj.filter}`;
        break;
      case 'filter-tag':
        path = makePathTag(obj.filter, obj.tag);
        break;
      case 'account':
        path = makePathAccount(obj.account);
        break;
      case 'post':
        path = makePathEntry(obj.cat, obj.author, obj.permlink);
        break;
      case 'sps':
        path = `/sps`;
        break;
      case 'witnesses':
        path = `/witnesses`;
        break;
      default:
        path = null;
        break;
    }

    if (path) {
      history.push(path);
    }
  };

  render() {
    return null;
  }
}

DeepLinkHandler.propTypes = {
  history: PropTypes.shape({}).isRequired
};

export default DeepLinkHandler;
