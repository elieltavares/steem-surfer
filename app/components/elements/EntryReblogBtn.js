/*
eslint-disable react/no-multi-comp
*/

import React, { Component } from 'react';

import PropTypes from 'prop-types';

import { FormattedMessage, FormattedRelative } from 'react-intl';

import { Popconfirm, Popover, message, Table, Modal, Badge } from 'antd';

import Tooltip from '../common/Tooltip';

import LoginRequired, { checkLogin } from '../helpers/LoginRequired';

import AccountLink from '../helpers/AccountLink';

import authorReputation from '../../utils/author-reputation';
import formatChainError from '../../utils/format-chain-error';

import { getItem, setItem } from '../../helpers/storage';

import { getAccounts, reblog } from '../../backend/steem-client';
import {
  getPostReblogCount,
  getPostReblogs
} from '../../backend/esteem-client';

class ReBloggersModal extends Component {
  render() {
    const { reblogs, onCancel, afterClose } = this.props;

    const modalTableColumns = [
      {
        dataIndex: 'account',
        key: 'account',
        render: (text, record) => (
          <span>
            <AccountLink {...this.props} username={text}>
              <span style={{ cursor: 'pointer', fontWeight: 'bold' }}>
                {text}
              </span>
            </AccountLink>{' '}
            <Badge
              count={record.reputation}
              style={{
                backgroundColor: '#fff',
                color: '#999',
                boxShadow: '0 0 0 1px #d9d9d9 inset'
              }}
            />
          </span>
        )
      },
      {
        dataIndex: 'timestamp',
        key: 'timestamp',
        width: 220,
        render: text => (
          <span title={text}>
            <FormattedRelative value={text} />
          </span>
        )
      }
    ];

    return (
      <Modal
        visible
        onCancel={onCancel}
        afterClose={afterClose}
        footer={false}
        width="80%"
        centered
        showHeader={false}
        title={
          <FormattedMessage
            id="entry-reblog.modal-title"
            values={{ n: reblogs.length }}
          />
        }
      >
        <Table
          dataSource={reblogs}
          columns={modalTableColumns}
          scroll={{ y: 310 }}
          showHeader={false}
          rowKey="account"
        />
      </Modal>
    );
  }
}

ReBloggersModal.defaultProps = {
  onCancel: () => {},
  afterClose: () => {}
};

ReBloggersModal.propTypes = {
  reblogs: PropTypes.arrayOf(Object).isRequired,
  onCancel: PropTypes.func,
  afterClose: PropTypes.func,
  intl: PropTypes.instanceOf(Object).isRequired
};

class EntryReblogBtn extends Component {
  constructor(props) {
    super(props);

    this.state = {
      processing: false,
      reblogs: [],
      reblogCount: 0,
      reblogModal: false
    };
  }

  componentDidMount() {
    this.mounted = true;

    this.fetchReblogStuff();
  }

  componentWillUnmount() {
    this.mounted = false;
  }

  mounted = false;

  stateSet = (obj, cb = undefined) => {
    if (this.mounted) {
      this.setState(obj, cb);
    }
  };

  fetchReblogStuff = () => {
    const { showDetail } = this.props;

    this.fetchReblogCount();

    if (showDetail) {
      this.fetchReblogs();
    }
  };

  fetchReblogCount = () => {
    const { entry } = this.props;

    return getPostReblogCount(entry.author, entry.permlink).then(r => {
      this.stateSet({ reblogCount: r });
      return r;
    });
  };

  fetchReblogs = () => {
    const { entry } = this.props;

    return getPostReblogs(entry.author, entry.permlink)
      .then(reblogs => {
        // Fetch account data from blockchain in order to get reputation.
        const ac = reblogs.map(a => a.account);
        return getAccounts(ac).then(accounts =>
          reblogs.map(v => {
            const account = accounts.find(a => a.name === v.account);
            if (account) {
              return Object.assign({}, v, {
                reputation: authorReputation(account.reputation)
              });
            }
            return v;
          })
        );
      })
      .then(reblogs =>
        reblogs.sort((a, b) => {
          // Sort data by reputation
          const keyA = a.reputation;
          const keyB = b.reputation;

          if (keyA > keyB) return -1;
          if (keyA < keyB) return 1;
          return 0;
        })
      )
      .then(reblogs => this.stateSet({ reblogs }));
  };

  countClicked = () => {
    this.stateSet({ reblogModal: true });
  };

  markAsReblogged = () => {
    const { activeAccount, entry } = this.props;
    const { author, permlink } = entry;

    setItem(`${activeAccount.username}-${author}-${permlink}-reblogged`, 1);
  };

  isReblogged = () => {
    const { activeAccount, entry } = this.props;

    const { author, permlink } = entry;

    if (
      getItem(`${activeAccount.username}-${author}-${permlink}-reblogged`) === 1
    ) {
      return true;
    }

    if (!activeAccount.accountData) {
      return false;
    }

    const { blog } = activeAccount.accountData;

    if (
      blog &&
      blog.some(e => e.author === author && e.permlink === permlink)
    ) {
      return true;
    }

    return false;
  };

  doReblog = () => {
    const { activeAccount, global, entry, intl } = this.props;

    this.stateSet({ processing: true });

    return reblog(activeAccount, global.pin, entry.author, entry.permlink)
      .then(resp => {
        message.success(intl.formatMessage({ id: 'entry-reblog.reblogged' }));
        this.markAsReblogged();
        this.fetchReblogStuff();
        return resp;
      })
      .catch(e => {
        message.error(formatChainError(e));
        if (
          e.jse_shortmsg &&
          String(e.jse_shortmsg).indexOf('has already reblogged')
        ) {
          // already blogged
          this.markAsReblogged();
        }
      })
      .finally(() => {
        this.stateSet({ processing: false });
      });
  };

  render() {
    const { intl, entry, activeAccount, showDetail } = this.props;
    const { reblogs, reblogCount, reblogModal } = this.state;

    const aElems = [];

    if (reblogCount > 0) {
      const btnProps = {};
      if (showDetail && reblogs.length > 0) {
        btnProps.onClick = this.countClicked;
      }

      const btn = (
        <span
          role="none"
          key="count-btn"
          className="reblog-count"
          {...btnProps}
        >
          {reblogCount}
        </span>
      );

      if (reblogs.length > 0 && !reblogModal) {
        const accounts = reblogs.map(x => `@${x.account}`).slice(0, 3);
        const moreCount = reblogs.length - 3;
        let content = accounts.join(', ');

        if (moreCount > 0) {
          content = `${content} ${intl.formatMessage(
            { id: 'entry-reblog.more-reblogs' },
            { n: moreCount }
          )}`;
        }

        aElems.push(
          <Popover
            key="reblog-popover"
            content={content}
            trigger="hover"
            placement="right"
          >
            {btn}
          </Popover>
        );
      } else {
        aElems.push(btn);
      }
    }

    if (reblogModal) {
      aElems.push(
        <ReBloggersModal
          key="modal"
          {...this.props}
          reblogs={reblogs}
          onCancel={() => this.stateSet({ reblogModal: false })}
        />
      );
    }

    if (!activeAccount) {
      return (
        <div className="reblog-btn">
          <LoginRequired {...this.props} requiredKeys={['posting']}>
            <a className="inner-btn">
              <Tooltip
                title={intl.formatMessage({ id: 'entry-reblog.reblog' })}
                mouseEnterDelay={2}
              >
                <i className="mi">repeat</i>
              </Tooltip>
            </a>
          </LoginRequired>

          {aElems}
        </div>
      );
    }

    const ownEntry =
      activeAccount !== null && activeAccount.username !== entry.author;
    if (!ownEntry) {
      return null;
    }

    const { processing } = this.state;

    const reblogged = this.isReblogged();
    const isLoginOk = checkLogin(activeAccount, ['posting']);
    const btnCls = `reblog-btn${reblogged ? ' reblogged-btn' : ''}${
      processing ? ' processing' : ''
    }`;

    if (!isLoginOk) {
      return (
        <div className={btnCls}>
          <LoginRequired {...this.props} requiredKeys={['posting']}>
            <a className="inner-btn">
              <Tooltip
                title={intl.formatMessage({ id: 'entry-reblog.reblog' })}
                mouseEnterDelay={2}
              >
                <i className="mi">repeat</i>
              </Tooltip>
            </a>
          </LoginRequired>

          {aElems}
        </div>
      );
    }

    return (
      <div className={btnCls}>
        <a className="inner-btn">
          <Popconfirm
            title={intl.formatMessage({ id: 'g.are-you-sure' })}
            placement="right"
            onConfirm={() => {
              this.doReblog();
            }}
          >
            <Tooltip
              title={intl.formatMessage({ id: 'entry-reblog.reblog' })}
              mouseEnterDelay={2}
            >
              <i className="mi">repeat</i>
            </Tooltip>
          </Popconfirm>
        </a>

        {aElems}
      </div>
    );
  }
}

EntryReblogBtn.defaultProps = {
  activeAccount: null,
  showDetail: false
};

EntryReblogBtn.propTypes = {
  global: PropTypes.shape({
    pin: PropTypes.string
  }).isRequired,
  entry: PropTypes.shape({
    author: PropTypes.string.isRequired,
    permlink: PropTypes.string.isRequired
  }).isRequired,
  activeAccount: PropTypes.instanceOf(Object),
  showDetail: PropTypes.bool,
  intl: PropTypes.instanceOf(Object).isRequired
};

export default EntryReblogBtn;
