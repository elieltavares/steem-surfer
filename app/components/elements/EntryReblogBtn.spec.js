/* eslint-disable */

import React from 'react';
import Enzyme from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import EntryReblogBtn from './EntryReblogBtn';
import { mountWithIntl } from '../../helpers/intl-test';

Enzyme.configure({ adapter: new Adapter() });

describe('EntryReblogBtn', () => {
  it('(1) Render', () => {
    const props = {
      global: {},
      entry: {
        author: 'foo',
        permlink: 'bar-baz'
      }
    };
    const wrapper = mountWithIntl(<EntryReblogBtn {...props} />);
    expect(wrapper.html()).toMatchSnapshot();
  });
});
