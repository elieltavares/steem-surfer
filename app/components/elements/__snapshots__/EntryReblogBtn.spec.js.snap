// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EntryReblogBtn (1) Render 1`] = `"<div class=\\"reblog-btn\\"><a class=\\"inner-btn\\"><i class=\\"mi tooltip-handler\\">repeat</i></a></div>"`;
